dask/array/core.py

```diff
@@ -1726,19 +1724,38 @@ def common_blockdim(blockdims):
     --------
 
     >>> common_blockdim([(3,), (2, 1)])
-    set([(2, 1)])
+    (2, 1)
+    >>> common_blockdim([(1, 2), (2, 1)])
+    (1, 1, 1)
     >>> common_blockdim([(2, 2), (3, 1)])  # doctest: +SKIP
     Traceback (most recent call last):
         ...
     ValueError: Chunks do not align
     """
     non_trivial_dims = set([d for d in blockdims if len(d) > 1])
-    if len(non_trivial_dims) > 1:
-        raise ValueError('Chunks do not align %s' % non_trivial_dims)
-    elif non_trivial_dims:
-        return non_trivial_dims
-    else:
-        return blockdims
+    if len(non_trivial_dims) == 1:
+        return first(non_trivial_dims)
+    if len(non_trivial_dims) == 0:
+        return max(blockdims, key=first)
+
+    if len(set(map(sum, non_trivial_dims))) > 1:
+        raise ValueError("Chunks do not add up to same value", blockdims)
+
+    chunks = list(map(list, non_trivial_dims))
+    total = sum(first(non_trivial_dims))
+    i = 0
+
+    out = []
+    while i < total:
+        m = min(c[0] for c in chunks)
+        out.append(m)
+        for c in chunks:
+            c[0] -= m
+            if c[0] == 0:
+                c.pop(0)
```

Comment:
Popping the first item in a list is slow. This is probably going to result in painful performance issues with a moderately large number of chunks.

I would either use numpy arrays (replace the view), collections.deque or a separate integer index pointing to the first value.